app/src/forms/form/service.js

```diff
@@ -1040,6 +1043,114 @@ const service = {
       throw error;
     }
   },
+
+  // -----------------------------------------------------------------------------
+  // External API
+  // -----------------------------------------------------------------------------
+
+  listExternalAPIs: (formId) => {
+    return ExternalAPI.query().modify('filterFormId', formId);
+  },
+
+  listExternalAPIAlgorithms: () => {
+    return Object.values(ENCRYPTION_ALGORITHMS).map((x) => ({
+      code: x,
+      display: x,
+    }));
+  },
+
+  listExternalAPIStatusCodes: () => {
+    return ExternalAPIStatusCode.query();
+  },
+
+  validateExternalAPI: (data) => {
+    if (!data) {
+      throw new Problem(422, `'externalAPI record' cannot be empty.`);
+    }
+    if (data.sendApiKey) {
+      if (!data.apiKeyHeader || !data.apiKey) {
+        throw new Problem(422, `'apiKeyHeader' and 'apiKey' are required when 'sendApiKey' is true.`);
+      }
+    }
+    if (data.sendUserToken) {
+      if (!data.userTokenHeader) {
+        throw new Problem(422, `'userTokenHeader' is required when 'sendUserToken' is true.`);
+      }
+    }
+    if (data.sendUserInfo) {
+      if (data.userInfoEncrypted && !data.userInfoHeader) {
+        throw new Problem(422, `'userInfoHeader' is required when 'sendUserInfo' and 'userInfoEncrypted' are true.`);
+      }
+      if (data.userInfoEncrypted) {
+        if (!Object.values(ENCRYPTION_ALGORITHMS).includes(data.userInfoEncryptionAlgo)) {
+          throw new Problem(422, `'${data.userInfoEncryptionAlgo}' is not a valid Encryption Algorithm.`);
+        }
+        if (!data.userInfoEncryptionKey) {
+          throw new Problem(422, `'userInfoEncryptionKey' is required when 'userInfoEncrypted' is true.`);
+        }
+      }
+    }
+  },
+
+  createExternalAPI: async (formId, data, currentUser) => {
+    service.validateExternalAPI(data);
+
+    let trx;
+    try {
+      trx = await ExternalAPI.startTransaction();
+      data.id = uuidv4();
+      // set status to SUBMITTED
+      data.code = ExternalAPIStatuses.SUBMITTED;
+      await ExternalAPI.query(trx).insert({
+        ...data,
+        createdBy: currentUser.usernameIdp,
+      });
+
+      await trx.commit();
```

Comment:
I see this pattern a lot in the services (and I've followed it too!) but I'm starting to wonder about it. Does this transaction `trx` actually do anything? We're only doing a single `insert`, so if it fails then there is nothing to roll back, is there?

Reply:
so removing transaction works in runtime... however, the unit tests will be pretty useless? at least with the transaction we could test for commit being called and rollback.  let me know if you want it removed and the unit tests updated.

Reply:
I think it would be better to remove the transactions (unless needed for multiple database changes) and have simple tests, unless we gain something in the code by having the transaction.

Reply:
sounds good. will remove where it is a single table. we may lose a couple of unit tests (I'm not too fond of tests that are 100% mocked). off for a couple of days but will pick this up ASAP.

Reply:
Agreed, the unit tests are usually useless for one-line service functions.